dask/base.py

```diff
@@ -132,14 +132,42 @@ def normalize_function(func):
 normalize_token.register(object,
         lambda a: normalize_function(a) if callable(a) else a)
 normalize_token.register(dict, lambda a: tuple(sorted(a.items())))
+
 with ignoring(ImportError):
     import pandas as pd
-    normalize_token.register(pd.DataFrame,
-            lambda a: (id(a), len(a), list(a.columns)))
-    normalize_token.register(pd.Series, lambda a: (id(a), len(a), a.name))
+
+    @partial(normalize_token.register, pd.Index)
+    def normalize_index(ind):
+        return tokenize(ind.name, ind.values)
+
+    @partial(normalize_token.register, pd.Categorical)
+    def normalize_categorical(cat):
+        return tokenize(cat.codes, cat.categories, cat.ordered)
+
+    @partial(normalize_token.register, pd.Series)
+    def normalize_series(s):
+        return tokenize(s.name, s._data.blocks[0].values, s.dtype, s.index)
```

Comment:
str and hash are mostly free

```python
In [1]: from hashlib import md5

In [2]: %time _ = md5(str(['1y9hfshfs90f8hfs0shf', 'hf90s8hfsh90h098hf3']))
CPU times: user 9 µs, sys: 2 µs, total: 11 µs
Wall time: 11.9 µs

In [3]: %timeit md5(str(['1y9hfshfs90f8hfs0shf', 'hf90s8hfsh90h098hf3']))
The slowest run took 20.88 times longer than the fastest. This could mean that an intermediate result is being cached 
1000000 loops, best of 3: 1.05 µs per loop
```